[yaml-cpp] Support dynamic library on linux

vmiheer opened the pull request:
**This prevents unixy compilers erroring out after seeing __declspec(dllimport|export)**
This is temporary change, ideally it should go upstream.

- Which triplets are supported/not supported? Have you updated the CI baseline?
linux-x64-dynamic, I don't think linux-x64-dynamic is official triplet so likely no CI impact.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes.

vmiheer committed:
[yaml-cpp] Support dynamic library on linux

This prevents unixy compilers erroring out after seeing __declspec(dllimport|export)
This is temporary change, ideally it should go upstream.

NancyLi1013 commented on ports/yaml-cpp/portfile.cmake:
```diff
@@ -14,6 +14,16 @@ else()
+    vcpkg_apply_patches(
+       SOURCE_PATH ${SOURCE_PATH}
+       PATCHES ${patchForLinux})
+endif()
```
Can you make a patch directly based on the current source instead of downloading from upstream?

vmiheer replied:
Like putting patch in vcpkg repository?

vmiheer replied:
@NancyLi1013, updated port version in CONTROL. I am not sure about https://github.com/microsoft/vcpkg/pull/15626#discussion_r556996841

NancyLi1013 replied:
Yes, putting patch in vcpkg repository.

vmiheer replied:
I think the patch is not related to the library support being compatible with vcpkg the patch deserves to be in upstream repository.

NancyLi1013 replied:
Your consideration is reasonable. But we recommend to use the patch provided by vcpkg to avoid the patch being applied failed due to some external reasons.

NancyLi1013 reviewed (commented):
Could you please also bump CONTROL file?

`Add Port-Version: 1`

vmiheer committed:
Bump port version in CONTROL

NancyLi1013 commented on ports/yaml-cpp/CONTROL:
```diff
@@ -2,3 +2,4 @@ Source: yaml-cpp
 Version: 0.6.3
```
```suggestion
Version: 0.6.3
Port-Version: 1
```

NancyLi1013 commented on ports/yaml-cpp/CONTROL:
```diff
@@ -2,3 +2,4 @@ Source: yaml-cpp
 Version: 0.6.3
 Homepage: https://github.com/jbeder/yaml-cpp
 Description: yaml-cpp is a YAML parser and emitter in C++ matching the YAML 1.2 spec.
+Port-Version: 1
```
```suggestion
```

NancyLi1013 committed:
Update ports/yaml-cpp/CONTROL

NancyLi1013 committed:
Update ports/yaml-cpp/CONTROL

NancyLi1013 commented:
Hi @vmiheer 

Could you please merge master to this PR?

Please also run the command `vcpkg x-add-version yaml-cpp`.

vmiheer committed:
1. Add local patchfile

2. Run vcpkg x-add-version yaml-cpp

vmiheer commented:
All done

NancyLi1013 commented:
@vmiheer 

Could you please merge master to this PR?

vmiheer committed:
Merge remote-tracking branch 'origin/master' into patch-5

vmiheer commented:
Done

NancyLi1013 commented:
/azp run

NancyLi1013 commented on ports/yaml-cpp/portfile.cmake:
```diff
@@ -14,6 +14,12 @@ else()
     set(YAML_BUILD_SHARED_LIBS OFF)
 endif()
```
Could you please update these codes as follows?

`string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" YAML_BUILD_SHARED_LIBS)`

Please see the doc https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md

NancyLi1013 commented on ports/yaml-cpp/portfile.cmake:
```diff
@@ -14,6 +14,12 @@ else()
     set(YAML_BUILD_SHARED_LIBS OFF)
 endif()
 
+if (NOT(WIN32) AND ${YAML_BUILD_SHARED_LIBS})
+    vcpkg_apply_patches(
+       SOURCE_PATH ${SOURCE_PATH}
+       PATCHES jbeder_yaml-cpp_958.patch)
+endif()
```
```suggestion
if (NOT WIN32 AND YAML_BUILD_SHARED_LIBS)
    vcpkg_apply_patches(
       SOURCE_PATH ${SOURCE_PATH}
       PATCHES jbeder_yaml-cpp_958.patch)
endif()
```
Seems that `jbeder_yaml-cpp_958.patch` is missing in this PR. So I wonder where do you get it now?

NancyLi1013 commented:
@vmiheer 

Could you please address the review suggestions?

NancyLi1013 commented:
Is work still being done for this PR? @vmiheer

vmiheer committed:
Merge branch 'master' into patch-5

vmiheer committed:
Add the patch file to repo

vmiheer commented:
Sorry for late reply, I was busy with other things. I have rebased the branch on master and added the patch file in the folder. I tested locally and seem to be fine.

NancyLi1013 commented:
You also need to address the review suggestions @vmiheer.

vmiheer commented:
I have added: https://github.com/vmiheer/vcpkg/commit/85bf14c032a3a53e4f144ba4a16c9d85c7d80445

vmiheer committed:
Update ports/yaml-cpp/portfile.cmake

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

NancyLi1013 commented:
If you don't have enough time to continue working on this PR now, you can convert this PR to draft until all work have been done  on your sides. @vmiheer
